### PR TITLE
add previewOffset option to useDrag

### LINKED ISSUE #5409: useDrag set x/y offset for setDragImage
### Provide a general summary of the feature here

Right now the useDrag hook provides no way to customize the position of the DragPreview. I tried a variety of ways in CSS to try to position the preview myself by was unsuccessful. 

### 🤔 Expected Behavior?

I should have some way to set the x/y

### 😯 Current Behavior

This is not way to do this

### 💁 Possible Solution

I think what's needed is some extra options for the useDrag hook. Maybe something along the lines of the following

```ts
interface UseDragOptions {
  previewOffset?: { x: number; y: number } | 'center';
}
```

### 🔦 Context

I have a DragPreview that I can't position. If the above API looks like something you would merge I'd happily make the PR. If the API isn't quite right I'm also down to implement whatever would land!

### 💻 Examples

_No response_

### 🧢 Your Company/Team

Descript

### 🕷 Tracking Issue

_No response_

### PR DESCRIPTION
**Pull request by hipstersmoothie**
Closes #5409

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [x] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

Go to the story I added and drag.

https://github.com/adobe/react-spectrum/assets/1192452/ebe0358e-cc59-463b-9a92-9de5d495edb7

### CODE REVIEW

**Review by snowystinger — APPROVED** [review]
works as described, approving and hopefully we'll get an API discussion going soon

File: packages/@react-aria/dnd/src/useDrag.ts [hunk 1]
```diff
@@ -137,9 +139,20 @@ export function useDrag(options: DragOptions): DragResult {
...
+          y = offset.y;
+        } else if (
+          options.previewOffset === 'center' || 
+          (!options.previewOffset && (x > size.width || y > size.height))
+        ) {
           x = size.width / 2;
           y = size.height / 2;
+        } else if (options.previewOffset) {
+          x = options.previewOffset.x;
+          y = options.previewOffset.y;
```
Comment (LFDanLu) [inline packages/@react-aria/dnd/src/useDrag.ts:155]: When testing this locally I noticed that if you set the offset larger number the drag preview will initially render with the proper offset but then slide itself to still be intersecting with the mouse. This seems to be matching native behavior as per this modified example from MDN: https://codepen.io/dl1644/pen/LYqaWYV but still feels pretty weird. 

Perhaps we should clamp the provided x,y value so the element is always intersecting the pointer?

**Review by LFDanLu — COMMENTED** [review]
I think the api is fine, would only be nice to get this into the hooks that use useDrag like useDraggableItem and useDragAndDrop so they can also specify the preview offset (out of scope for this PR of course). Will need to wait for consensus from the team before merging

**Comment by snowystinger** [comment]
Hey, thank you for your patience! We finally got an API discussion on this. We'd like to change it so that the value can be changed if the drag preview changes. Such as if you have multiple Drag Previews for multiple data types.

We will need to update this in two places.
https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/dnd/src/useDragAndDrop.ts#L70
and
https://github.com/adobe/react-spectrum/blob/d330a9dc13dd8376f5d2f8cdffcd7bce53a8fbf4/packages/%40react-aria/dnd/src/DragPreview.tsx#L18

The return type of both of these should be altered to be
```
JSX.Element | {element: JSX.Element, x:number, y:number}
```

Below is a rough sketch of the steps to implement it, there may be some holes.

Then in `DragPreview` we'd change the useImperativeHandler to something along these lines
```
useImperativeHandle(ref, () => (items: DragItem[], callback: (node: HTMLElement, x?: number, y?: number) => void) => {
    // This will be called during the onDragStart event by useDrag. We need to render the
    // preview synchronously before this event returns so we can call event.dataTransfer.setDragImage.
    let result = render(items);
    flushSync(() => {
      setChildren(result.element);
    });

    // Yield back to useDrag to set the drag image.
    callback(domRef.current, result.x, result.y);
    ...
```

And in https://github.com/adobe/react-spectrum/blob/d330a9dc13dd8376f5d2f8cdffcd7bce53a8fbf4/packages/%40react-aria/dnd/stories/dnd.stories.tsx#L449 we can use it roughly like so

```
          return {
            element: (
              <div className={classNames(dndStyles, 'draggable', 'is-drag-preview', {'is-dragging-multiple': selectedKeys.size > 1})}>
                ...
              </div>
            ),
            x: 50,
            y: 50
          }
```

And then in useDrag https://github.com/adobe/react-spectrum/blob/d330a9dc13dd8376f5d2f8cdffcd7bce53a8fbf4/packages/%40react-aria/dnd/src/useDrag.ts#L132
```
    if (typeof options.preview?.current === 'function') {
      // console.log('options', options.preview.current)
      options.preview.current(items, (node, xUser, yUser) => {
```

**Comment by snowystinger** [comment]
Going to close this one for now. We can reopen if it's picked up again or someone else can make a new PR

**Pull request closed without merging**